spec/ParseUser.spec.js

```diff
@@ -13,92 +13,40 @@ const passwordCrypto = require('../lib/password');
- 
-     const provider = getMockFacebookProvider();
-     Parse.User._registerAuthenticationProvider(provider);
-     const user = await Parse.User._logInWith('facebook');
-     await Parse.User.logOut();
-     //Parse.User._registerAuthenticationProvider(provider);
-     await Parse.User._logInWith('facebook');
- 
- 
-   });
-   const getMockFacebookProviderWithIdToken = function (id, token) {
-     return {
-       authData: {
-         id: id,
-         access_token: token,
-         expiration_date: new Date().toJSON(),
-       },
-       shouldError: false,
-       loggedOut: false,
-       synchronizedUserId: null,
-       synchronizedAuthToken: null,
-       synchronizedExpiration: null,
- 
-       authenticate: function (options) {
-         if (this.shouldError) {
-           options.error(this, 'An error occurred');
-         } else if (this.shouldCancel) {
-           options.error(this, null);
-         } else {
-           options.success(this, this.authData);
-         }
-       },
-       restoreAuthentication: function (authData) {
-         if (!authData) {
-           this.synchronizedUserId = null;
-           this.synchronizedAuthToken = null;
-           this.synchronizedExpiration = null;
-           return true;
-         }
-         this.synchronizedUserId = authData.id;
-         this.synchronizedAuthToken = authData.access_token;
-         this.synchronizedExpiration = authData.expiration_date;
-         return true;
-       },
-       getAuthType: function () {
-         return 'facebook';
-       },
-       deauthenticate: function () {
-         this.loggedOut = true;
-         this.restoreAuthentication(null);
-       },
-     };
-   };
-  const getMockFacebookProvider = function () {
-     return getMockFacebookProviderWithIdToken('8675309', 'jenny');
-   };
- 
- });
- 
 describe('Parse.User testing', () => {
+  it('retrieves original object when user logs in with third party auth', async done => {
+    let counter = 0;
+    Parse.Cloud.beforeSave(Parse.User, async request => {
+      const { object: user } = request;
+      if (user.get('authData') != null && user.isNew()) {
+        console.log('Before save: new user');
+        user.setUsername('facebookaccount');
+      }
+    });
+    Parse.Cloud.afterSave(Parse.User, async request => {
+      if (!request.original) {
+        console.log('After save: new user ' + counter);
+        const user = request.object;
+        if (user.get('authData') != null) {
+          if (counter > 0) {
+            done.fail();
+          }
+          counter++;
+          console.log(
+            'Counter: ' + counter + ' User signed up using auth -> username ' + user.get('username')
+          );
+        }
+      } else {
+        console.log('User updated profile');
+      }
+    });
+
+    const provider = getMockFacebookProvider();
+    Parse.User._registerAuthenticationProvider(provider);
+    await Parse.User.logOut();
+    await Parse.User._logInWith('facebook');
```

Comment:
`_logInWith` is a private method, why did you not use [linkWith](https://docs.parseplatform.org/js/guide/#signing-up-and-logging-in)?

Reply:
Because the purpose of the test is just to demonstrate where the issue is, the issue occurs even when you use the android sdk, also the link method is used to link an already registered account, in this case, I have tested it only when a user performs a signup/login. even if I use what you proposed I get the error 
`Unhandled promise rejection: TypeError: Parse.User._linkWith is not a function`

Reply:
> I get the error `Unhandled promise rejection: TypeError: Parse.User._linkWith is not a function`

Can you try `Parse.User.linkWith`?

> the link method is used to link an already registered account

It should be for both, new sign up and login with existing account.

> the issue occurs even when you use the android sdk

Let's try to reproduce it without Android SDK, so we know it's a server issue not an SDK issue.

Reply:
same error for linkWith
TypeError: Parse.User.linkWith is not a function